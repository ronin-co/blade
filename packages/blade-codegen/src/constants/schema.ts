import { SyntaxKind, factory } from 'typescript';

import { identifiers } from '@/src/constants/identifiers';

import type { TypeNode } from 'typescript';

import type { ModelField } from '@/src/types/model';
import type { DML_QUERY_TYPES } from 'blade-compiler';

/**
 * A list of all model field types & their TypeScript type mapping.
 */
export const MODEL_TYPE_TO_SYNTAX_KIND_KEYWORD = {
  blob: factory.createTypeReferenceNode(identifiers.compiler.storedObject),
  boolean: factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword),
  date: factory.createTypeReferenceNode(identifiers.primitive.date),
  json: factory.createUnionTypeNode([
    factory.createTypeReferenceNode(identifiers.utils.jsonObject),
    factory.createTypeReferenceNode(identifiers.utils.jsonArray),
  ]),
  link: factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword),
  number: factory.createKeywordTypeNode(SyntaxKind.NumberKeyword),
  string: factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
} satisfies Record<ModelField['type'], TypeNode>;

/**
 * A simple object mapping all DML query types to their human readable string.
 */
export const READABLE_DML_QUERY_TYPES = {
  add: 'Add',
  count: 'Count',
  get: 'Get',
  remove: 'Remove',
  set: 'Set',
  use: 'Get',
} satisfies Record<(typeof DML_QUERY_TYPES)[number] | 'use', string>;

/**
 * A list of all default field slugs for RONIN models.
 *
 * This is designed to match the properties as part of `ResultRecord`.
 */
export const DEFAULT_FIELD_SLUGS = [
  'id',
  'ronin.createdAt',
  'ronin.createdBy',
  'ronin.updatedAt',
  'ronin.updatedBy',
] satisfies Array<string>;
