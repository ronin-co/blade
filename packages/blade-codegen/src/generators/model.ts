import { SyntaxKind, factory } from 'typescript';

import { identifiers, typeArgumentIdentifiers } from '@/src/constants/identifiers';
import { DEFAULT_FIELD_SLUGS } from '@/src/constants/schema';
import { triggerOptionsInterface } from '@/src/declarations';
import { convertToPascalCase } from '@/src/utils/slug';
import { mapRoninFieldToTypeNode, remapNestedFields } from '@/src/utils/types';

import type { TypeAliasDeclaration, TypeParameterDeclaration } from 'typescript';

import type { ModelField } from '@/src/types/model';
import type { PopulatedModel } from 'blade-compiler';

/**
 * Generate all required type definitions for a provided schema model.
 *
 * The plural model type will be mapped to an array of the singular model type and
 * extended with the plural model properties.
 *
 * @param models - All models of the addressed space.
 *
 * @returns An array of type nodes to be added to the `index.d.ts` file.
 */
export const generateModelTypes = (
  models: Array<PopulatedModel>,
): Array<TypeAliasDeclaration> => {
  const nodes = new Array<TypeAliasDeclaration>(triggerOptionsInterface);

  for (const model of models) {
    const fields: Array<ModelField> = Object.entries(model.fields)
      .map(([slug, field]) => ({ ...field, slug }) as ModelField)
      .filter((field) => !DEFAULT_FIELD_SLUGS.includes(field.slug));

    const modelIdentifier = {
      singular: factory.createIdentifier(convertToPascalCase(model.slug)),
      plural: factory.createIdentifier(convertToPascalCase(model.pluralSlug)),
    };

    const hasLinkFields = fields.some(
      (field) =>
        field.type === 'link' && models.some((model) => model.slug === field.target),
    );

    const modelInterfaceTypeParameters = new Array<TypeParameterDeclaration>();
    const linkFieldKeys = fields
      .filter((field) => field.type === 'link')
      .map((field) =>
        factory.createLiteralTypeNode(factory.createStringLiteral(field.slug)),
      );

    /**
     * ```ts
     * <U extends Array<...> | 'all' = []>
     * ```
     */
    const usingGenericDec = factory.createTypeParameterDeclaration(
      undefined,
      typeArgumentIdentifiers.using,
      factory.createUnionTypeNode([
        factory.createTypeReferenceNode(identifiers.primitive.array, [
          factory.createUnionTypeNode(linkFieldKeys),
        ]),
        factory.createLiteralTypeNode(factory.createStringLiteral('all')),
      ]),
      factory.createTupleTypeNode([]),
    );

    if (hasLinkFields) modelInterfaceTypeParameters.push(usingGenericDec);

    /**
     * ```ts
     * SchemaSlugSchema<U>
     * ```
     */
    const modelSchemaName = factory.createTypeReferenceNode(
      modelIdentifier.singular,
      hasLinkFields
        ? [factory.createTypeReferenceNode(typeArgumentIdentifiers.using)]
        : [],
    );

    /**
     * ```ts
     * export type SchemaSlug<U extends Array<...> | 'all' = []> = ResultRecord & {
     *  // ...
     * };
     * ```
     */
    const singularModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      modelIdentifier.singular,
      modelInterfaceTypeParameters,
      factory.createIntersectionTypeNode([
        factory.createTypeReferenceNode(
          factory.createQualifiedName(
            identifiers.namespace.utils.name,
            identifiers.namespace.utils.resultRecord,
          ),
        ),
        factory.createTypeLiteralNode(
          remapNestedFields(fields)
            .sort(([a], [b]) => a.localeCompare(b))
            .map(([fieldSlug, field]) => {
              if (Array.isArray(field)) {
                const sortedFields = field.sort((a, b) => a.slug.localeCompare(b.slug));
                return factory.createPropertySignature(
                  undefined,
                  fieldSlug,
                  undefined,
                  factory.createTypeLiteralNode(
                    sortedFields.map((nestedField) =>
                      factory.createPropertySignature(
                        undefined,
                        nestedField.slug,
                        undefined,
                        factory.createUnionTypeNode(
                          mapRoninFieldToTypeNode(nestedField, models),
                        ),
                      ),
                    ),
                  ),
                );
              }

              return factory.createPropertySignature(
                undefined,
                fieldSlug,
                undefined,
                factory.createUnionTypeNode(mapRoninFieldToTypeNode(field, models)),
              );
            }),
        ),
      ]),
    );

    /**
     * ```ts
     * Array<SchemaSlug>;
     * ```
     */
    const pluralModelArrayTypeDec = factory.createTypeReferenceNode(
      identifiers.primitive.array,
      [modelSchemaName],
    );

    /**
     * ```ts
     * {
     *  moreBefore?: string;
     *  moreAfter?: string;
     * };
     * ```
     */
    const pluralModelPaginationPropsTypeDec = factory.createTypeLiteralNode([
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier('moreBefore'),
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier('moreAfter'),
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
    ]);

    /**
     * ```ts
     * export type SchemaPluralSlug<U extends Array<...> | 'all' = []> = Array<SchemaSlug> & {
     *  moreBefore?: string;
     *  moreAfter?: string;
     * };
     * ```
     */
    const pluralModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      modelIdentifier.plural,
      modelInterfaceTypeParameters,
      factory.createIntersectionTypeNode([
        pluralModelArrayTypeDec,
        pluralModelPaginationPropsTypeDec,
      ]),
    );

    nodes.push(singularModelTypeDec, pluralModelTypeDec);
  }

  return nodes;
};
