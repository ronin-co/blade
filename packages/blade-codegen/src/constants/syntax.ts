/**
 * A list of all inferred combined instruction properties for RONIN models.
 */
export const INFERRED_COMBINED_INSTRUCTION_PROPERTIES = [
  'after',
  'before',
  'including',
  'limitedTo',
  'orderedBy',
  'selecting',

  // TODO(@nurodev): Move out & only include `using` if the model includes any link fields.
  'using',
] satisfies Array<string>;
