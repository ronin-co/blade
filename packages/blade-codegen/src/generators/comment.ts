import { READABLE_DML_QUERY_TYPES } from '@/src/constants/schema';

import type { QueryType } from '@/src/types/query';
import type { PopulatedModel } from 'blade-compiler';

interface GenerateQueryTypeCommentResult {
  singular: string;
  plural: string;
}

/**
 * Generate a text comment for a provided model & a given query type.
 *
 * @param modelName - The name of the model to generate a comment for.
 * @param queryType - The query type to generate a comment for.
 *
 * @returns An object containing both the singular and plural comment strings.
 */
export const generateQueryTypeComment = (
  model: PopulatedModel,
  queryType: QueryType,
): GenerateQueryTypeCommentResult => ({
  singular: `* ${READABLE_DML_QUERY_TYPES[queryType]} a single ${model.name ?? model.slug} record `,
  plural: `* ${READABLE_DML_QUERY_TYPES[queryType]} multiple ${model.name ?? model.slug} records `,
});
