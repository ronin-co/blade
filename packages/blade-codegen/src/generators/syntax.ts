import { factory } from 'typescript';

import { identifiers, typeArgumentIdentifiers } from '@/src/constants/identifiers';
import { DEFAULT_FIELD_SLUGS } from '@/src/constants/schema';
import { sharedQueryOptionsParameter } from '@/src/declarations';
import { convertToPascalCase } from '@/src/utils/slug';

import type {
  Identifier,
  IntersectionTypeNode,
  TypeAliasDeclaration,
  TypeNode,
} from 'typescript';

import type { Model } from '@/src/types/model';

/**
 * @todo(@nurodev): Add documentation
 */
export const generateUsingSyntax = (
  model: Model,
  modelNode: TypeNode,
  promise: boolean,
  plural: boolean,
): IntersectionTypeNode => {
  const hasLinkFields = Object.values(model.fields).some(
    (field) => field.type === 'link',
  );
  if (!hasLinkFields)
    return factory.createIntersectionTypeNode([
      factory.createExpressionWithTypeArguments(
        identifiers.blade.reducedFunction,
        undefined,
      ),
      factory.createFunctionTypeNode(
        [
          factory.createTypeParameterDeclaration(
            undefined,
            typeArgumentIdentifiers.default,
            undefined,
            modelNode,
          ),
        ],
        [
          factory.createParameterDeclaration(
            undefined,
            undefined,
            'value',
            undefined,
            factory.createIndexedAccessTypeNode(
              factory.createTypeReferenceNode(identifiers.compiler.combinedInstructions),
              factory.createLiteralTypeNode(factory.createStringLiteral('using')),
            ),
          ),
        ],
        promise
          ? factory.createTypeReferenceNode(identifiers.primitive.promise, [
              factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
            ])
          : factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
      ),
    ]);

  /**
   * @todo(@nurodev): Add documentation
   */
  const arrayFieldsType = factory.createUnionTypeNode([
    factory.createTypeReferenceNode(identifiers.primitive.array, [
      factory.createUnionTypeNode(
        Object.entries(model.fields)
          .filter(([, field]) => field.type === 'link')
          .map(([name]) =>
            factory.createLiteralTypeNode(factory.createStringLiteral(name)),
          ),
      ),
    ]),
    factory.createLiteralTypeNode(factory.createStringLiteral('all')),
  ]);

  /**
   * @todo(@nurodev): Add documentation
   */
  const baseModelWithFields = factory.createTypeReferenceNode(
    factory.createIdentifier(convertToPascalCase(model.slug)),
    [factory.createTypeReferenceNode(typeArgumentIdentifiers.using)],
  );
  const modelNodeWithFields = promise
    ? factory.createTypeReferenceNode(identifiers.primitive.promise, [
        baseModelWithFields,
      ])
    : baseModelWithFields;

  /**
   * @todo(@nurodev): Add documentation
   */
  const inferredCallSignature = factory.createCallSignature(
    [
      factory.createTypeParameterDeclaration(
        undefined,
        typeArgumentIdentifiers.using,
        arrayFieldsType,
      ),
    ],
    [
      factory.createParameterDeclaration(
        undefined,
        undefined,
        'fields',
        undefined,
        factory.createTypeReferenceNode(typeArgumentIdentifiers.using),
      ),
    ],
    plural
      ? modelNodeWithFields
      : factory.createUnionTypeNode([
          modelNodeWithFields,
          factory.createLiteralTypeNode(factory.createNull()),
        ]),
  );

  /**
   * @todo(@nurodev): Add documentation
   */
  const overrideCallSignaure = factory.createCallSignature(
    [
      factory.createTypeParameterDeclaration(
        undefined,
        typeArgumentIdentifiers.default,
        undefined,
        modelNode,
      ),
    ],
    [
      factory.createParameterDeclaration(
        undefined,
        undefined,
        'fields',
        undefined,
        arrayFieldsType,
      ),
    ],
    promise
      ? factory.createTypeReferenceNode(identifiers.primitive.promise, [
          factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
        ])
      : factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
  );

  return factory.createIntersectionTypeNode([
    factory.createExpressionWithTypeArguments(
      identifiers.blade.reducedFunction,
      undefined,
    ),
    factory.createTypeLiteralNode([inferredCallSignature, overrideCallSignaure]),
  ]);
};

/**
 * @todo(@nurodev): Add documentation
 */
export const generateWithSyntax = (
  name: string | Identifier,
  modelNode: TypeNode,
  model: Model,
  promise: boolean,
): TypeAliasDeclaration => {
  const modelUserFieldEntries = Object.entries(model.fields).filter(
    ([slug]) => !DEFAULT_FIELD_SLUGS.some((field) => field.includes(slug)),
  );

  return factory.createTypeAliasDeclaration(
    undefined,
    name,
    undefined,
    factory.createIntersectionTypeNode([
      factory.createTypeReferenceNode(
        factory.createQualifiedName(identifiers.namespace.utils.name, name),
        [modelNode],
      ),
      factory.createTypeLiteralNode(
        modelUserFieldEntries.map(([slug, field]) => {
          /**
           * @todo(@nurodev): Add documentation
           */
          const baseFieldProperty = factory.createIndexedAccessTypeNode(
            factory.createTypeReferenceNode(convertToPascalCase(model.slug)),
            factory.createLiteralTypeNode(factory.createStringLiteral(slug)),
          );

          const parameterNodes = new Array<TypeNode>(baseFieldProperty);
          if (field.type === 'link') {
            /**
             * @todo(@nurodev): Add documentation
             */
            const nestedFieldProperty = factory.createTypeReferenceNode(
              identifiers.primitive.partial,
              [
                factory.createIndexedAccessTypeNode(
                  factory.createTypeReferenceNode(convertToPascalCase(model.slug), [
                    factory.createTupleTypeNode([
                      factory.createLiteralTypeNode(factory.createStringLiteral(slug)),
                    ]),
                  ]),
                  factory.createLiteralTypeNode(factory.createStringLiteral(slug)),
                ),
              ],
            );
            parameterNodes.push(nestedFieldProperty);
          }

          return factory.createPropertySignature(
            undefined,
            slug,
            undefined,
            factory.createFunctionTypeNode(
              [
                factory.createTypeParameterDeclaration(
                  undefined,
                  typeArgumentIdentifiers.default,
                  undefined,
                  modelNode,
                ),
              ],
              [
                factory.createParameterDeclaration(
                  undefined,
                  undefined,
                  slug,
                  undefined,
                  factory.createUnionTypeNode(parameterNodes),
                ),
                sharedQueryOptionsParameter,
              ],
              promise
                ? factory.createTypeReferenceNode(identifiers.primitive.promise, [
                    factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
                  ])
                : factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
            ),
          );
        }),
      ),
    ]),
  );
};
