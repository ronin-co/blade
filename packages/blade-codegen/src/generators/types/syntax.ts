import { SyntaxKind, factory } from 'typescript';

import { identifiers, typeArgumentIdentifiers } from '@/src/constants/identifiers';
import { DEFAULT_FIELD_SLUGS } from '@/src/constants/schema';
import { INFERRED_COMBINED_INSTRUCTION_PROPERTIES } from '@/src/constants/syntax';
import { convertToPascalCase } from '@/src/utils/slug';
import { mapRoninFieldToTypeNode } from '@/src/utils/types';

import type { TypeAliasDeclaration, TypeElement, TypeNode } from 'typescript';

import type { Model, ModelField } from '@/src/types/model';

import type {} from 'typescript';

/**
 * TODO(@nurodev): Add documentation
 *
 * @example
 * ```ts
 * type UserSyntax<S> = ReducedFunction & {
 *  // ...
 * };
 * ```
 *
 * @param models - All models of the addressed space.
 */
export const generateModelSyntaxTypes = (
  models: Array<Model>,
): Array<TypeAliasDeclaration> => {
  const nodes = new Array<TypeAliasDeclaration>();

  const schemaTypeArgumentNode = factory.createTypeReferenceNode(
    typeArgumentIdentifiers.schema,
  );

  for (const model of models) {
    const modelIdentifier = factory.createIdentifier(
      `${convertToPascalCase(model.slug)}Syntax`,
    );

    /**
     * ```ts
     * <T = S>(options?: Partial<CombinedInstructions>): T;
     * ```
     */
    const rootInstructionCallSignature = factory.createCallSignature(
      [
        factory.createTypeParameterDeclaration(
          undefined,
          typeArgumentIdentifiers.default,
          undefined,
          schemaTypeArgumentNode,
        ),
      ],
      [
        factory.createParameterDeclaration(
          undefined,
          undefined,
          'options',
          factory.createToken(SyntaxKind.QuestionToken),
          factory.createTypeReferenceNode(identifiers.primitive.partial, [
            factory.createTypeReferenceNode(
              identifiers.compiler.combinedInstructions,
              undefined,
            ),
          ]),
          undefined,
        ),
      ],
      factory.createTypeReferenceNode(typeArgumentIdentifiers.default, undefined),
    );

    const literalMembers = new Array<TypeElement>(rootInstructionCallSignature);

    for (const propertyName of INFERRED_COMBINED_INSTRUCTION_PROPERTIES) {
      /**
       * after: <T = S>(value: CombinedInstructions["after"]) => T;
       * before: <T = S>(value: CombinedInstructions["before"]) => T;
       * including: <T = S>(value: CombinedInstructions["including"]) => T;
       * limitedTo: <T = S>(value: CombinedInstructions["limitedTo"]) => T;
       * orderedBy: <T = S>(value: CombinedInstructions["orderedBy"]) => T;
       * selecting: <T = S>(value: CombinedInstructions["selecting"]) => T;
       * using: <T = S>(value: CombinedInstructions["using"]) => T;
       */
      const memberSignature = factory.createPropertySignature(
        undefined,
        propertyName,
        undefined,
        factory.createFunctionTypeNode(
          [
            factory.createTypeParameterDeclaration(
              undefined,
              typeArgumentIdentifiers.default,
              undefined,
              schemaTypeArgumentNode,
            ),
          ],
          [
            factory.createParameterDeclaration(
              undefined,
              undefined,
              'value',
              undefined,
              factory.createIndexedAccessTypeNode(
                factory.createTypeReferenceNode(
                  identifiers.compiler.combinedInstructions,
                  undefined,
                ),
                factory.createLiteralTypeNode(factory.createStringLiteral(propertyName)),
              ),
              undefined,
            ),
          ],
          factory.createTypeReferenceNode(typeArgumentIdentifiers.default, undefined),
        ),
      );

      literalMembers.push(memberSignature);
    }

    /**
     * ```ts
     * with: ReducedFunction & {
     *  <T = S>(options?: Partial<CombinedInstructions>): T;
     *  id: <T = S>(value: ResultRecord["id"]) => T;
     *  ronin: ReducedFunction & {
     *    createdAt: <T = S>(value: ResultRecord["ronin"]["createdAt"]) => T;
     *    createdBy: <T = S>(value: ResultRecord["ronin"]["createdBy"]) => T;
     *    updatedAt: <T = S>(value: ResultRecord["ronin"]["updatedAt"]) => T;
     *    updatedBy: <T = S>(value: ResultRecord["ronin"]["updatedBy"]) => T;
     *  };
     * };
     * ```
     */
    const withPropertySignature = factory.createPropertySignature(
      undefined,
      'with',
      undefined,
      factory.createIntersectionTypeNode([
        factory.createExpressionWithTypeArguments(
          identifiers.utils.reducedFunction,
          undefined,
        ),
        factory.createTypeLiteralNode([
          rootInstructionCallSignature,
          ...getWithPropertySignatureMembers(
            models,
            model.fields,
            schemaTypeArgumentNode,
          ),
        ]),
      ]),
    );
    literalMembers.push(withPropertySignature);

    nodes.push(
      factory.createTypeAliasDeclaration(
        undefined,
        modelIdentifier,
        [
          factory.createTypeParameterDeclaration(
            undefined,
            typeArgumentIdentifiers.schema,
            undefined,
            undefined,
          ),
        ],
        factory.createIntersectionTypeNode([
          factory.createExpressionWithTypeArguments(
            identifiers.utils.reducedFunction,
            undefined,
          ),
          factory.createTypeLiteralNode(literalMembers),
        ]),
      ),
    );
  }

  return nodes;
};

/**
 * @todo(@nurodev): Add documentation
 */
const getWithPropertySignatureMembers = (
  models: Array<Model>,
  fields: Model['fields'],
  schemaTypeArgumentNode: TypeNode,
): Array<TypeElement> => {
  const members = new Array<TypeElement>();

  // Add all top-level RONIN fields, such as `id`.
  const topLevelFields = DEFAULT_FIELD_SLUGS.filter((slug) => !slug.startsWith('ronin.'));
  for (const slug of topLevelFields) {
    members.push(
      factory.createPropertySignature(
        undefined,
        slug,
        undefined,
        factory.createFunctionTypeNode(
          [
            factory.createTypeParameterDeclaration(
              undefined,
              typeArgumentIdentifiers.default,
              undefined,
              schemaTypeArgumentNode,
            ),
          ],
          [
            factory.createParameterDeclaration(
              undefined,
              undefined,
              'value',
              undefined,
              factory.createIndexedAccessTypeNode(
                factory.createTypeReferenceNode(
                  identifiers.syntax.resultRecord,
                  undefined,
                ),
                factory.createLiteralTypeNode(factory.createStringLiteral(slug)),
              ),
            ),
          ],
          factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
        ),
      ),
    );
  }

  // Add all nested RONIN meta fields, such as `ronin.createdAt`, `ronin.createdBy`, etc.
  const metaFields = DEFAULT_FIELD_SLUGS.filter((slug) => slug.startsWith('ronin.'));
  const metaFieldsPropertySignatures = metaFields.map((slug) => {
    const normalizedSlug = slug.replaceAll('ronin.', '');

    return factory.createPropertySignature(
      undefined,
      normalizedSlug,
      undefined,
      factory.createFunctionTypeNode(
        [
          factory.createTypeParameterDeclaration(
            undefined,
            typeArgumentIdentifiers.default,
            undefined,
            schemaTypeArgumentNode,
          ),
        ],
        [
          factory.createParameterDeclaration(
            undefined,
            undefined,
            'value',
            undefined,
            factory.createIndexedAccessTypeNode(
              factory.createIndexedAccessTypeNode(
                factory.createTypeReferenceNode(identifiers.syntax.resultRecord),
                factory.createLiteralTypeNode(factory.createStringLiteral('ronin')),
              ),
              factory.createLiteralTypeNode(factory.createStringLiteral(normalizedSlug)),
            ),
          ),
        ],
        factory.createTypeReferenceNode(typeArgumentIdentifiers.default),
      ),
    );
  });
  members.push(
    factory.createPropertySignature(
      undefined,
      'ronin',
      undefined,
      factory.createIntersectionTypeNode([
        factory.createExpressionWithTypeArguments(
          identifiers.utils.reducedFunction,
          undefined,
        ),
        factory.createTypeLiteralNode(metaFieldsPropertySignatures),
      ]),
    ),
  );

  // Create property signatures for all model fields.
  for (const [slug, field] of Object.entries(fields)) {
    if (DEFAULT_FIELD_SLUGS.some((field) => field.includes(slug))) continue;

    members.push(
      factory.createPropertySignature(
        undefined,
        slug,
        undefined,
        factory.createFunctionTypeNode(
          [
            factory.createTypeParameterDeclaration(
              undefined,
              typeArgumentIdentifiers.default,
              undefined,
              schemaTypeArgumentNode,
            ),
          ],
          [
            factory.createParameterDeclaration(
              undefined,
              undefined,
              'value',
              undefined,
              factory.createUnionTypeNode(
                mapRoninFieldToTypeNode(field as ModelField, models),
              ),
              undefined,
            ),
          ],
          factory.createTypeReferenceNode(typeArgumentIdentifiers.default, undefined),
        ),
      ),
    );
  }

  return members;
};
