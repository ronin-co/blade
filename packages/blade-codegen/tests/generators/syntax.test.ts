import { describe, expect, test } from 'bun:test';
import { link, model, string } from 'blade-syntax/schema';
import { factory } from 'typescript';

import { generateUsingSyntax, generateWithSyntax } from '@/src/generators/syntax';
import { printNodes } from '@/src/utils/print';
import { convertToPascalCase } from '@/src/utils/slug';

import type { PopulatedModel as Model } from 'blade-compiler';

describe('syntax', () => {
  describe('using', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: string(),
      },
    }) as unknown as Model;
    const SingularAccountModelNode = factory.createUnionTypeNode([
      factory.createTypeReferenceNode(convertToPascalCase(AccountModel.slug)),
      factory.createLiteralTypeNode(factory.createNull()),
    ]);
    const PluralAccountModelNode = factory.createArrayTypeNode(
      factory.createTypeReferenceNode(convertToPascalCase(AccountModel.slug)),
    );

    const PostModel = model({
      slug: 'post',
      pluralSlug: 'posts',
      fields: {
        title: string(),
        author: link({ target: 'account' }),
      },
    }) as unknown as Model;
    const SingularPostModelNode = factory.createUnionTypeNode([
      factory.createTypeReferenceNode(convertToPascalCase(PostModel.slug)),
      factory.createLiteralTypeNode(factory.createNull()),
    ]);
    const PluralPostModelNode = factory.createArrayTypeNode(
      factory.createTypeReferenceNode(convertToPascalCase(PostModel.slug)),
    );

    describe('singular', () => {
      describe('synchronous', () => {
        test('with no link fields', () => {
          const typesResult = generateUsingSyntax(
            AccountModel,
            SingularAccountModelNode,
            false,
            false,
          );

          const typesResultStr = printNodes([typesResult]);

          expect(typesResultStr).toMatchSnapshot();
        });

        test('with a link field', () => {
          const typesResult = generateUsingSyntax(
            PostModel,
            SingularPostModelNode,
            false,
            false,
          );

          const typesResultStr = printNodes([typesResult]);

          expect(typesResultStr).toMatchSnapshot();
        });
      });
      describe('asynchronous', () => {
        test('with no link fields', () => {
          const typesResult = generateUsingSyntax(
            AccountModel,
            SingularAccountModelNode,
            true,
            false,
          );

          const typesResultStr = printNodes([typesResult]);

          expect(typesResultStr).toMatchSnapshot();
        });

        test('with a link field', () => {
          const typesResult = generateUsingSyntax(
            PostModel,
            SingularPostModelNode,
            true,
            false,
          );

          const typesResultStr = printNodes([typesResult]);

          expect(typesResultStr).toMatchSnapshot();
        });
      });
    });

    describe('plural', () => {
      describe('synchronous', () => {
        test('with no link fields', () => {
          const typesResult = generateUsingSyntax(
            AccountModel,
            PluralAccountModelNode,
            false,
            true,
          );

          const typesResultStr = printNodes([typesResult]);

          expect(typesResultStr).toMatchSnapshot();
        });

        test('with a link field', () => {
          const typesResult = generateUsingSyntax(
            PostModel,
            PluralPostModelNode,
            false,
            true,
          );

          const typesResultStr = printNodes([typesResult]);

          expect(typesResultStr).toMatchSnapshot();
        });
      });
      describe('asynchronous', () => {
        test('with no link fields', () => {
          const typesResult = generateUsingSyntax(
            AccountModel,
            PluralAccountModelNode,
            true,
            true,
          );

          const typesResultStr = printNodes([typesResult]);

          expect(typesResultStr).toMatchSnapshot();
        });

        test('with a link field', () => {
          const typesResult = generateUsingSyntax(
            PostModel,
            PluralPostModelNode,
            true,
            true,
          );

          const typesResultStr = printNodes([typesResult]);

          expect(typesResultStr).toMatchSnapshot();
        });
      });
    });
  });

  describe('with', () => {
    const AccountModel = model({
      slug: 'account',
      pluralSlug: 'accounts',
      fields: {
        name: string(),
        email: string({ required: true, unique: true }),
      },
    }) as unknown as Model;
    const SingularAccountModelNode = factory.createUnionTypeNode([
      factory.createTypeReferenceNode(convertToPascalCase(AccountModel.slug)),
      factory.createLiteralTypeNode(factory.createNull()),
    ]);
    const PluralAccountModelNode = factory.createArrayTypeNode(
      factory.createTypeReferenceNode(convertToPascalCase(AccountModel.slug)),
    );

    describe('singular', () => {
      test('synchronous', () => {
        const typesResult = generateWithSyntax(
          AccountModel,
          SingularAccountModelNode,
          false,
        );

        const typesResultStr = printNodes([typesResult]);

        expect(typesResultStr).toMatchSnapshot();
      });
      test('asynchronous', () => {
        const typesResult = generateWithSyntax(
          AccountModel,
          SingularAccountModelNode,
          true,
        );

        const typesResultStr = printNodes([typesResult]);

        expect(typesResultStr).toMatchSnapshot();
      });
    });

    describe('plural', () => {
      test('synchronous', () => {
        const typesResult = generateWithSyntax(
          AccountModel,
          PluralAccountModelNode,
          false,
        );

        const typesResultStr = printNodes([typesResult]);

        expect(typesResultStr).toMatchSnapshot();
      });
      test('asynchronous', () => {
        const typesResult = generateWithSyntax(
          AccountModel,
          PluralAccountModelNode,
          true,
        );

        const typesResultStr = printNodes([typesResult]);

        expect(typesResultStr).toMatchSnapshot();
      });
    });
  });
});
