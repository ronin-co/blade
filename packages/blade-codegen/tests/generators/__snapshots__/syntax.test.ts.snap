// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`syntax using singular synchronous with no link fields 1`] = `
"ReducedFunction & (<T = Account | null>(value: CombinedInstructions["using"]) => T)
"
`;

exports[`syntax using singular synchronous with a link field 1`] = `
"ReducedFunction & {
    <U extends Array<"author"> | "all">(fields: U): Post<U> | null;
    <T = Post | null>(fields: Array<"author"> | "all"): T;
}
"
`;

exports[`syntax using singular asynchronous with no link fields 1`] = `
"ReducedFunction & (<T = Account | null>(value: CombinedInstructions["using"]) => Promise<T>)
"
`;

exports[`syntax using singular asynchronous with a link field 1`] = `
"ReducedFunction & {
    <U extends Array<"author"> | "all">(fields: U): Promise<Post<U>> | null;
    <T = Post | null>(fields: Array<"author"> | "all"): Promise<T>;
}
"
`;

exports[`syntax using plural synchronous with no link fields 1`] = `
"ReducedFunction & (<T = Account[]>(value: CombinedInstructions["using"]) => T)
"
`;

exports[`syntax using plural synchronous with a link field 1`] = `
"ReducedFunction & {
    <U extends Array<"author"> | "all">(fields: U): Post<U>;
    <T = Post[]>(fields: Array<"author"> | "all"): T;
}
"
`;

exports[`syntax using plural asynchronous with no link fields 1`] = `
"ReducedFunction & (<T = Account[]>(value: CombinedInstructions["using"]) => Promise<T>)
"
`;

exports[`syntax using plural asynchronous with a link field 1`] = `
"ReducedFunction & {
    <U extends Array<"author"> | "all">(fields: U): Promise<Post<U>>;
    <T = Post[]>(fields: Array<"author"> | "all"): Promise<T>;
}
"
`;

exports[`syntax with singular synchronous 1`] = `
"type WithQuery = Utils.WithQuery<Account | null> & {
    name: <T = Account | null>(name: Account["name"], options?: Record<string, unknown>) => T;
    email: <T = Account | null>(email: Account["email"], options?: Record<string, unknown>) => T;
};
"
`;

exports[`syntax with singular asynchronous 1`] = `
"type WithQueryPromise = Utils.WithQueryPromise<Account | null> & {
    name: <T = Account | null>(name: Account["name"], options?: Record<string, unknown>) => Promise<T>;
    email: <T = Account | null>(email: Account["email"], options?: Record<string, unknown>) => Promise<T>;
};
"
`;

exports[`syntax with plural synchronous 1`] = `
"type WithQuery = Utils.WithQuery<Account[]> & {
    name: <T = Account[]>(name: Account["name"], options?: Record<string, unknown>) => T;
    email: <T = Account[]>(email: Account["email"], options?: Record<string, unknown>) => T;
};
"
`;

exports[`syntax with plural asynchronous 1`] = `
"type WithQueryPromise = Utils.WithQueryPromise<Account[]> & {
    name: <T = Account[]>(name: Account["name"], options?: Record<string, unknown>) => Promise<T>;
    email: <T = Account[]>(email: Account["email"], options?: Record<string, unknown>) => Promise<T>;
};
"
`;
